// @superset-ui/core/esm/query/buildQueryObject.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/* eslint-disable camelcase */
import processGroupby from './processGroupby';
import convertMetric from './convertMetric';
import processFilters from './processFilters';
import extractExtras from './extractExtras';
import extractQueryFields from './extractQueryFields';
import { appendExtraFormData, overrideExtraFormData } from './processExtraFormData';
export const DTTM_ALIAS = '__timestamp';
/**
 * Build the common segments of all query objects (e.g. the granularity field derived from
 * either sql alchemy or druid). The segments specific to each viz type is constructed in the
 * buildQuery method for each viz type (see `wordcloud/buildQuery.ts` for an example).
 * Note the type of the formData argument passed in here is the type of the formData for a
 * specific viz, which is a subtype of the generic formData shared among all viz types.
 */

export default function buildQueryObject(formData) {
  const {
    annotation_layers = [],
    extra_form_data = {},
    time_range,
    since,
    until,
    order_desc,
    row_limit,
    row_offset,
    limit,
    timeseries_limit_metric,
    queryFields,
    granularity,
    url_params = {}
  } = formData,
        residualFormData = _objectWithoutPropertiesLoose(formData, ["annotation_layers", "extra_form_data", "time_range", "since", "until", "order_desc", "row_limit", "row_offset", "limit", "timeseries_limit_metric", "queryFields", "granularity", "url_params"]);

  const {
    append_form_data = {},
    override_form_data = {}
  } = extra_form_data;
  const numericRowLimit = Number(row_limit);
  const numericRowOffset = Number(row_offset);
  const {
    metrics,
    groupby,
    columns
  } = extractQueryFields(residualFormData, queryFields);
  const groupbySet = new Set([...columns, ...groupby]);
  const extras = extractExtras(formData);
  const extrasAndfilters = processFilters(_extends({}, formData, {}, extras));

  let queryObject = _extends({
    time_range,
    since,
    until,
    granularity
  }, extras, {}, extrasAndfilters, {
    annotation_layers,
    groupby: processGroupby(Array.from(groupbySet)),
    is_timeseries: groupbySet.has(DTTM_ALIAS),
    metrics: metrics.map(convertMetric),
    order_desc: typeof order_desc === 'undefined' ? true : order_desc,
    orderby: [],
    row_limit: row_limit == null || Number.isNaN(numericRowLimit) ? undefined : numericRowLimit,
    row_offset: row_offset == null || Number.isNaN(numericRowOffset) ? undefined : numericRowOffset,
    timeseries_limit: limit ? Number(limit) : 0,
    timeseries_limit_metric: timeseries_limit_metric ? convertMetric(timeseries_limit_metric) : null,
    url_params
  }); // append and override extra form data used by native filters


  queryObject = appendExtraFormData(queryObject, append_form_data);
  queryObject = overrideExtraFormData(queryObject, override_form_data);
  return queryObject;
}
