// @superset-ui/core/lib/query/api/v1/makeApi.js
"use strict";

exports.__esModule = true;
exports.default = makeApi;

var _rison = _interopRequireDefault(require("rison"));

var _connection = require("../../../connection");

var _handleError = _interopRequireDefault(require("./handleError"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

const validRequestTypes = new Set(['form', 'json', 'search', 'rison']);

function isPayloadless(method) {
  return !method || method === 'GET' || method === 'DELETE' || method === 'HEAD';
}
/**
 * Generate an API caller with predefined configs/typing and consistent
 * return values.
 */


function makeApi(_ref) {
  let {
    endpoint,
    method,
    requestType: requestType_,
    responseType,
    processResponse
  } = _ref,
      requestOptions = _objectWithoutPropertiesLoose(_ref, ["endpoint", "method", "requestType", "responseType", "processResponse"]);

  // use `search` payload (searchParams) when it's a GET request
  const requestType = requestType_ || (isPayloadless(method) ? 'search' : 'json');

  if (!validRequestTypes.has(requestType)) {
    throw new Error("Invalid request payload type, choose from: " + [...validRequestTypes].join(' | '));
  }

  async function request(payload, {
    client = _connection.SupersetClient
  } = {
    client: _connection.SupersetClient
  }) {
    try {
      const requestConfig = _extends({}, requestOptions, {
        method,
        endpoint
      });

      if (requestType === 'search') {
        requestConfig.searchParams = payload;
      } else if (requestType === 'rison') {
        requestConfig.endpoint = endpoint + "?q=" + _rison.default.encode(payload);
      } else if (requestType === 'form') {
        requestConfig.postPayload = payload;
      } else {
        requestConfig.jsonPayload = payload;
      }

      let result;
      const response = await client.request(_extends({}, requestConfig, {
        parseMethod: 'raw'
      }));

      if (responseType === 'text') {
        result = await response.text();
      } else if (responseType === 'raw' || responseType === null) {
        result = response;
      } else {
        result = await response.json(); // if response json has an "error" field

        if (result && typeof result === 'object' && 'error' in result) {
          return (0, _handleError.default)(result);
        }
      }

      const typedResult = result;
      return processResponse ? processResponse(typedResult) : typedResult;
    } catch (error) {
      return (0, _handleError.default)(error);
    }
  }

  request.method = method;
  request.endpoint = endpoint;
  request.requestType = requestType;
  return request;
}
