// @superset-ui/core/esm/chart-composition/ChartFrame.js
import _pt from "prop-types";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import { isDefined } from '../utils';

function checkNumber(input) {
  return isDefined(input) && typeof input === 'number';
}

export default class ChartFrame extends PureComponent {
  render() {
    const {
      contentWidth,
      contentHeight,
      width,
      height,
      renderContent
    } = this.props;
    const overflowX = checkNumber(contentWidth) && contentWidth > width;
    const overflowY = checkNumber(contentHeight) && contentHeight > height;

    if (overflowX || overflowY) {
      return /*#__PURE__*/React.createElement("div", {
        style: {
          height,
          overflowX: overflowX ? 'auto' : 'hidden',
          overflowY: overflowY ? 'auto' : 'hidden',
          width
        }
      }, renderContent({
        height: Math.max(contentHeight != null ? contentHeight : 0, height),
        width: Math.max(contentWidth != null ? contentWidth : 0, width)
      }));
    }

    return renderContent({
      height,
      width
    });
  }

}

_defineProperty(ChartFrame, "propTypes", {
  contentWidth: _pt.number,
  contentHeight: _pt.number,
  height: _pt.number.isRequired,
  renderContent: _pt.func.isRequired,
  width: _pt.number.isRequired
});

_defineProperty(ChartFrame, "defaultProps", {
  renderContent() {}

});
